// react-underline-tabbar/lib/prop-types/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.defaultProps = exports.propTypes = undefined;

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _utils = require('../utils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var propTypes = exports.propTypes = {
  tabs: _propTypes2.default.array,
  pos: _propTypes2.default.object,
  width: _propTypes2.default.number,
  height: _propTypes2.default.number,
  goToPage: _utils.noop,
  activeTab: _propTypes2.default.number,
  renderTab: _propTypes2.default.oneOfType([_propTypes2.default.func, _propTypes2.default.element]),
  scrollPosition: _propTypes2.default.string,
  style: _propTypes2.default.object,
  scrollViewStyle: _propTypes2.default.object,
  underlineStyle: _propTypes2.default.object,
  tabStyle: _propTypes2.default.object,
  tabActiveStyle: _propTypes2.default.object,
  tabTextStyle: _propTypes2.default.object,
  tabTextActiveStyle: _propTypes2.default.object,
  hasUnderline: _propTypes2.default.bool,
  scrollEnabled: _propTypes2.default.bool,
  hasAnimation: _propTypes2.default.bool,
  duration: _propTypes2.default.number,
  vertical: _propTypes2.default.bool,
  isAutoSize: _propTypes2.default.bool
};

var defaultProps = exports.defaultProps = {
  tabs: [],
  pos: null,
  width: undefined,
  height: undefined,
  goToPage: _utils.noop,
  activeTab: 0,
  renderTab: null,
  scrollPosition: 'center',
  style: {},
  scrollViewStyle: {},
  underlineStyle: {},
  tabStyle: {},
  tabActiveStyle: {},
  tabTextStyle: {},
  tabTextActiveStyle: {},
  hasUnderline: true,
  scrollEnabled: true,
  hasAnimation: true,
  duration: 400,
  vertical: true,
  isAutoSize: false
};
